Rubberduck.CodeAnalysis/Inspections/Concrete/EmptyMethodInspection.cs

```diff
@@ -0,0 +1,131 @@
+    /// Identifies empty module member blocks.
+    /// </summary>
+    /// <why>
+    /// Methods containing no executable statements are misleading as they appear to be doing something which they actually don't.
+    /// This might be the result of delaying the actual implementation for a later stage of development, and then forgetting all about that.
+    /// </why>
+    /// <example hasResults="true">
+    /// <![CDATA[
+    /// Sub Foo()
+    ///     ' ...
+    /// End Sub
+    /// ]]>
+    /// </example>
+    /// <example hasResults="false">
+    /// <![CDATA[
+    /// Sub Foo()
+    ///     MsgBox "?"
+    /// End Sub
+    /// ]]>
+    /// </example>
+    [Experimental(nameof(ExperimentalNames.EmptyBlockInspections))]
+    internal class EmptyMethodInspection : InspectionBase// EmptyBlockInspectionBase
+    {
+        public EmptyMethodInspection(RubberduckParserState state)
+            : base(state) { }
+
+        protected override IEnumerable<IInspectionResult> DoGetInspectionResults()
+        {
+            // Exclude empty members in user interfaces, as long as all members of the interface are empty,
+            // since some VB users might use concrete user defined classes as interfaces,
+            // while RD marks them as interfaces all the same.
+
+            return UserDeclarations.OfType<ModuleBodyElementDeclaration>()
+                .Where(bodyElement => !BlockContainsExecutableStatements(bodyElement.Block))
+                .GroupBy(bodyElement => bodyElement.ComponentName)
+                // Exclude results from user interfaces
+                .Where(bodyElements => !State.DeclarationFinder.FindAllUserInterfaces()
+                                       // where all members of that interface contain no executables
+                                       .Where(interfaceModule => interfaceModule.ComponentName == bodyElements.Key
+                                                                && interfaceModule.Members.Count == bodyElements.Count())
+                                       .Any())
+                .SelectMany(bodyElements => bodyElements)
+                .Select(result => new DeclarationInspectionResult(this,
+                                                                  string.Format(InspectionResults.EmptyMethodInspection,
+                                                                                result.DeclarationType.ToFormatted(),
+                                                                                result.IdentifierName),
+                                                                  result));
+        }
+
+        private bool BlockContainsExecutableStatements(BlockContext block)
+        {
+            return block?.children != null && ContainsExecutableStatements(block.children);
+        }
+
+        private bool ContainsExecutableStatements(IList<Antlr4.Runtime.Tree.IParseTree> blockChildren)
+        {
+            foreach (var child in blockChildren)
```

Comment:
I'm wondering if this would be better expressed as a LINQ statement. Something like this untested aircode:
```
blockChildren.Any(child => 
  child is BlockStmtContext blockStmt && 
  blockStmt.mainBlockStmt().Any(mainBlockStmt => 
    !mainBlockStmt.GetChild(0) is VariableStmtContext &&
    !mainBlockStmt.GetChild(0) is ConstStmtContext));
```

Reply:
I tested this simplified version:
```
        private bool ContainsExecutableStatementsLinq(IList<IParseTree> blockChildren)
        {
            return blockChildren.Any(child => child is BlockStmtContext blockStmt
                                              && !(blockStmt.mainBlockStmt() == null
                                                  || IsConstantOrVariable(blockStmt.mainBlockStmt().GetChild(0))));
        }

        private bool IsConstantOrVariable(IParseTree context)
        {
            return context is VariableStmtContext || context is ConstStmtContext;
        }
```
which passed all tests I was able to throw on it, and still couldn't beat first implementation:
`Old - AVG = 0.000340821022727273, Total = 11.9969`
`Linq - AVG = 0.000448607954545455, Total = 15.791`

I don't think this is the right direction to choose.

Reply:
Yeah at first all inspections had LINQ everywhere - while not a "hot path" as hot as, say, parsing & resolving, inspections still need every bit of performance they can get; LINQ has its place elsewhere, but I agree a fully spelled-out `foreach` loop here is probably a better choice.